File: packages/esm-format-go/pkg/esm/simulation_fixtures_test.go
package esm

// Execution runner for inline `tests` blocks on the tests/simulation/*.esm
// physics fixtures (gt-l1fk). Mirrors the Julia reference at
// packages/EarthSciSerialization.jl/test/tests_blocks_execution_test.jl.
//
// The Go binding is parse-only (no ODE solver), so this runner does not
// execute assertions numerically. Instead it closes the schema-vs-binding gap
// for every tests/simulation/ fixture by:
//   1. Parsing the fixture via Load (schema-validated).
//   2. Walking every model/reaction_system that carries a `tests` block,
//      confirming each Test has a time_span and at least one assertion with
//      the required (variable, time, expected) shape.
//
// If a future Go ODE backend lands, this runner is the place to wire
// numerical execution in — the fixture walk and the tolerance-precedence
// helpers already live in this package (resolveUnitsFixtureTol).

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// checkTestsBlock validates the structural shape of one `tests` block entry
// and returns the number of assertions it declares.
func checkTestsBlock(t *testing.T, label string, testsRaw []interface{}) int {
	t.Helper()
	assertions := 0
	for i, tcRaw := range testsRaw {
		tc, ok := tcRaw.(map[string]interface{})
		if !ok {
			t.Fatalf("%s: tests[%d] not an object", label, i)
		}
		id, _ := tc["id"].(string)
		if id == "" {
			t.Fatalf("%s: tests[%d] missing id", label, i)
		}
		tsRaw, ok := tc["time_span"].(map[string]interface{})
		if !ok {
			t.Fatalf("%s/%s: missing time_span", label, id)
		}
		if _, ok := tsRaw["start"].(float64); !ok {
			t.Fatalf("%s/%s: time_span.start not numeric", label, id)
		}
		if _, ok := tsRaw["end"].(float64); !ok {
			t.Fatalf("%s/%s: time_span.end not numeric", label, id)
		}
		aRaw, _ := tc["assertions"].([]interface{})
		if len(aRaw) == 0 {
			t.Fatalf("%s/%s: expected at least one assertion", label, id)
		}
		for j, raw := range aRaw {
			a, ok := raw.(map[string]interface{})
			if !ok {
				t.Fatalf("%s/%s: assertions[%d] not an object", label, id, j)
			}
			if v, _ := a["variable"].(string); v == "" {
				t.Fatalf("%s/%s: assertions[%d] missing variable", label, id, j)
			}
			if _, ok := a["time"].(float64); !ok {
				t.Fatalf("%s/%s: assertions[%d].time not numeric", label, id, j)
			}
			if _, ok := a["expected"].(float64); !ok {
				t.Fatalf("%s/%s: assertions[%d].expected not numeric", label, id, j)
			}
			assertions++
		}
	}
	return assertions
}

// TestSimulationFixturesBlocksExecution walks tests/simulation/*.esm and
// verifies every inline `tests` block round-trips cleanly through the Go
// binding. This is the parse-only counterpart to the Julia runner at
// packages/EarthSciSerialization.jl/test/tests_blocks_execution_test.jl.
//
// When a Go ODE simulator lands, replace the `skipNumeric = true` branch
// with real execution and keep the fixture walk.
func TestSimulationFixturesBlocksExecution(t *testing.T) {
	repoRoot, err := filepath.Abs(filepath.Join("..", "..", "..", ".."))
	if err != nil {
		t.Fatalf("resolve repo root: %v", err)
	}
	simDir := filepath.Join(repoRoot, "tests", "simulation")
	entries, err := os.ReadDir(simDir)
	if err != nil {
		t.Fatalf("read %s: %v", simDir, err)
	}

	var names []string
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(e.Name(), ".esm") {
			names = append(names, e.Name())
		}
	}
	if len(names) == 0 {
		t.Fatalf("no .esm fixtures in %s", simDir)
	}

	totalTests := 0
	totalAssertions := 0
	for _, name := range names {
		t.Run(name, func(t *testing.T) {
			path := filepath.Join(simDir, name)

			// 1. Load via Go's schema-validated loader. We only need the
			//    side effect (schema + subsystem-ref resolution); the Go
			//    struct does not yet expose the `tests` block.
			if _, err := Load(path); err != nil {
				t.Fatalf("Load(%s): %v", name, err)
			}

			// 2. Walk raw JSON to find inline `tests` blocks. The Go struct
			//    does not yet expose them, so we walk the doc directly — same
			//    pattern used by units_fixture_test.go.
			raw, err := os.ReadFile(path)
			if err != nil {
				t.Fatalf("read %s: %v", path, err)
			}
			var doc map[string]interface{}
			if err := json.Unmarshal(raw, &doc); err != nil {
				t.Fatalf("unmarshal %s: %v", name, err)
			}

			fixtureTests := 0
			models := mapOr(doc, "models")
			for mname, mraw := range models {
				model, ok := mraw.(map[string]interface{})
				if !ok {
					continue
				}
				tsRaw, _ := model["tests"].([]interface{})
				if len(tsRaw) == 0 {
					continue
				}
				label := fmt.Sprintf("%s/models/%s", name, mname)
				totalAssertions += checkTestsBlock(t, label, tsRaw)
				fixtureTests += len(tsRaw)
			}
			rsys := mapOr(doc, "reaction_systems")
			for rsname, rraw := range rsys {
				rs, ok := rraw.(map[string]interface{})
				if !ok {
					continue
				}
				tsRaw, _ := rs["tests"].([]interface{})
				if len(tsRaw) == 0 {
					continue
				}
				label := fmt.Sprintf("%s/reaction_systems/%s", name, rsname)
				totalAssertions += checkTestsBlock(t, label, tsRaw)
				fixtureTests += len(tsRaw)
			}
			totalTests += fixtureTests
		})
	}

	// Spec §4.7 tests/simulation fixtures must carry inline tests blocks
	// for at least one model — this guards against silent regressions where
	// all tests blocks get stripped during a migration.
	if totalTests == 0 {
		t.Fatalf("expected at least one inline test across tests/simulation/ fixtures, got 0")
	}
	if totalAssertions == 0 {
		t.Fatalf("expected at least one assertion across tests/simulation/ fixtures, got 0")
	}
}

File: packages/earthsci-toolkit/src/simulation-fixtures.test.ts
/**
 * Execution runner for inline `tests` blocks on the tests/simulation/*.esm
 * physics fixtures (gt-l1fk). Mirrors the Julia reference at
 * packages/EarthSciSerialization.jl/test/tests_blocks_execution_test.jl.
 *
 * The TypeScript binding is parse-only (no ODE solver), so this runner does
 * not execute assertions numerically. Instead it closes the schema-vs-binding
 * gap by parsing every tests/simulation/ fixture, round-tripping it through
 * save -> load, and validating the shape of every inline `tests` block.
 *
 * When a JS-side ODE backend lands, this runner is the place to wire
 * numerical execution in — the fixture walk is already here.
 */

import { describe, it, expect } from 'vitest';
import { readdirSync, readFileSync } from 'fs';
import { join } from 'path';
import { load, save } from './index.js';

const simulationDir = join(__dirname, '../../../tests/simulation');

function findSimulationFixtures(): string[] {
  return readdirSync(simulationDir)
    .filter((name) => name.endsWith('.esm'))
    .sort();
}

function validateTestsBlock(label: string, tests: unknown): number {
  expect(Array.isArray(tests), `${label}: tests is not an array`).toBe(true);
  const arr = tests as any[];
  let assertionsSeen = 0;
  for (let i = 0; i < arr.length; i++) {
    const tc = arr[i];
    expect(tc, `${label}: tests[${i}] not an object`).toBeTypeOf('object');
    expect(typeof tc.id, `${label}: tests[${i}].id not string`).toBe('string');
    expect(tc.id.length > 0, `${label}/${i}: empty id`).toBe(true);

    expect(tc.time_span, `${label}/${tc.id}: missing time_span`).toBeTypeOf('object');
    expect(typeof tc.time_span.start, `${label}/${tc.id}: time_span.start not number`).toBe('number');
    expect(typeof tc.time_span.end, `${label}/${tc.id}: time_span.end not number`).toBe('number');

    expect(Array.isArray(tc.assertions), `${label}/${tc.id}: assertions not array`).toBe(true);
    expect(tc.assertions.length, `${label}/${tc.id}: expected >=1 assertion`).toBeGreaterThan(0);
    for (let j = 0; j < tc.assertions.length; j++) {
      const a = tc.assertions[j];
      expect(typeof a.variable, `${label}/${tc.id}/a${j}: variable not string`).toBe('string');
      expect(typeof a.time, `${label}/${tc.id}/a${j}: time not number`).toBe('number');
      expect(typeof a.expected, `${label}/${tc.id}/a${j}: expected not number`).toBe('number');
      assertionsSeen++;
    }
  }
  return assertionsSeen;
}

describe('tests/simulation/ inline tests blocks', () => {
  const fixtures = findSimulationFixtures();

  it('finds at least one simulation fixture', () => {
    expect(fixtures.length).toBeGreaterThan(0);
  });

  // Aggregate across fixtures to guard against a regression that drops all
  // tests blocks — each fixture individually may legitimately omit them.
  let globalTests = 0;
  let globalAssertions = 0;

  it.each(fixtures)('%s: parses and round-trips', (name) => {
    const path = join(simulationDir, name);
    const text = readFileSync(path, 'utf-8');

    const file = load(text);
    const serialized = save(file);
    const reloaded = load(serialized);
    expect(reloaded).toEqual(file);

    const raw = JSON.parse(text);
    let fixtureTests = 0;
    let fixtureAssertions = 0;

    const models = raw.models ?? {};
    for (const [mname, mraw] of Object.entries<any>(models)) {
      const tests = mraw?.tests;
      if (!tests || tests.length === 0) continue;
      fixtureAssertions += validateTestsBlock(`${name}/models/${mname}`, tests);
      fixtureTests += tests.length;
    }
    const rsys = raw.reaction_systems ?? {};
    for (const [rsname, rraw] of Object.entries<any>(rsys)) {
      const tests = rraw?.tests;
      if (!tests || tests.length === 0) continue;
      fixtureAssertions += validateTestsBlock(
        `${name}/reaction_systems/${rsname}`,
        tests,
      );
      fixtureTests += tests.length;
    }
    globalTests += fixtureTests;
    globalAssertions += fixtureAssertions;
  });

  // Spec §4.7 — tests/simulation must carry executable inline tests for at
  // least one component, otherwise a schema migration silently stripped them.
  it('aggregates at least one inline test across all fixtures', () => {
    expect(globalTests).toBeGreaterThan(0);
    expect(globalAssertions).toBeGreaterThan(0);
  });
});
